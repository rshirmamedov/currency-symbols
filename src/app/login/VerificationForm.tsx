"use client";
import { Button } from "@/components/Button";
import { Field } from "@/components/Field";
import Link from "next/link";
import { useForm } from "react-hook-form";
import { sleep } from "@/utils";
import { TIMEOUT } from "@/config";
import { validateVerificationCode } from "./validations";
import { useCallback } from "react";

type VerificationFieldValues = {
    code: string;
};

type VerificationFormProps = {
    onBackToLogin: () => void;
    onSubmit: () => void;
};

export const VerificationForm = (props: VerificationFormProps) => {
    const {
        register,
        handleSubmit,
        formState: { errors, isSubmitting },
    } = useForm<VerificationFieldValues>();

    const onSubmit = useCallback(async () => {
        await sleep(TIMEOUT);
        props.onSubmit();
    }, [props]);

    return (
        <div className="max-w-md w-full p-8 border border-gray-300 rounded-md bg-white">
            <h1 className="text-center text-2xl">Verification code</h1>
            <div>
                <form onSubmit={handleSubmit(onSubmit)}>
                    <Field
                        inputId="code"
                        type="number"
                        label="Verification code (4 digits)"
                        placeholder="Enter 4-digit verification code"
                        disabled={isSubmitting}
                        error={errors.code?.message}
                        {...register("code", { validate: validateVerificationCode })}
                    />
                    <Button
                        data-testid="verify-button"
                        title={isSubmitting ? "Verifying..." : "Verify"}
                        disabled={isSubmitting}
                    />
                </form>
            </div>
            <div className="text-center p-4 text-sm border">
                <span className="text-blue-600 hover:underline pr-8">
                    <Link href="" onClick={props.onBackToLogin}>
                        Back to login
                    </Link>
                </span>
                <span className="text-blue-600 hover:underline">
                    <Link href="/registration">Register</Link>
                </span>
            </div>
        </div>
    );
};
