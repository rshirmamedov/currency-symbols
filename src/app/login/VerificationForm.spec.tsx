import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { VerificationForm } from "./VerificationForm";

describe("VerificationForm", () => {
    it("renders required verification code error", async () => {
        render(<VerificationForm onBackToLogin={jest.fn()} onSubmit={jest.fn()} />);

        await userEvent.click(await screen.findByTestId("verify-button"));

        expect(await screen.findByTestId("code-field")).toMatchSnapshot();
    });

    it("renders invalid verification code error", async () => {
        const user = userEvent.setup();

        render(<VerificationForm onBackToLogin={jest.fn()} onSubmit={jest.fn()} />);

        const codeInput = screen.getByLabelText(/code/i);

        await user.type(codeInput, "12");

        await userEvent.click(await screen.findByTestId("verify-button"));

        expect(await screen.findByTestId("code-field")).toMatchSnapshot();
    });

    it("disables verify button while submitting", async () => {
        const user = userEvent.setup();

        render(<VerificationForm onBackToLogin={jest.fn()} onSubmit={jest.fn()} />);

        const codeInput = screen.getByLabelText(/code/i);

        await user.type(codeInput, "1234");

        await userEvent.click(await screen.findByTestId("verify-button"));

        expect(await screen.findByTestId("verify-button")).toMatchSnapshot();
    });
});
