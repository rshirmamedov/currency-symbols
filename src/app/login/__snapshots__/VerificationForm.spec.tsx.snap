// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerificationForm disables verify button while submitting 1`] = `
<button
  class="w-full py-2.5 px-4 text-sm rounded text-white bg-blue-600 hover:bg-blue-700 disabled:bg-blue-400 focus:outline-none mt-6"
  data-testid="verify-button"
  disabled=""
>
  Verifying...
</button>
`;

exports[`VerificationForm renders invalid verification code error 1`] = `
<div
  class="py-2"
  data-testid="code-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="code"
  >
    Verification code (4 digits)
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="code"
    name="code"
    placeholder="Enter 4-digit verification code"
    type="number"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Verification code is invalid
    </div>
  </div>
</div>
`;

exports[`VerificationForm renders required verification code error 1`] = `
<div
  class="py-2"
  data-testid="code-field"
>
  <label
    class="block text-sm font-medium leading-6 text-gray-900"
    for="code"
  >
    Verification code (4 digits)
  </label>
  <input
    class="w-full text-sm p-3 rounded outline-none border-2 focus:border-blue-500"
    id="code"
    name="code"
    placeholder="Enter 4-digit verification code"
    type="number"
  />
  <div
    class="pt-2"
  >
    <div
      class="text-sm text-red-600"
    >
      Verification code is required
    </div>
  </div>
</div>
`;
